File: server/src/utils/DocumentStats.ts
interface DocumentCounts {
  pubmed: number;
  semanticScholar: number;
  duplicates: number;
  selected: number;
  total: number;
}

class DocumentStats {
  private static projectStats: Map<string, DocumentCounts> = new Map();

  private static seededRandom(initialSeed: number): () => number {
    let seed = initialSeed;
    return () => {
      seed = (seed * 16807) % 2147483647;
      return (seed - 1) / 2147483646;
    };
  }

  static generateInitialStats(projectId: string): DocumentCounts {
    // If stats already exist for this project, return them
    if (this.projectStats.has(projectId)) {
      return this.projectStats.get(projectId)!;
    }

    // Create seeded random number generator
    const initialSeed = parseInt(projectId.replace(/\D/g, '')) || Date.now();
    const random = this.seededRandom(initialSeed);

    // Calculate total papers (800-1500)
    const total = Math.floor(random() * (1500 - 800 + 1)) + 800;

    // Split between PubMed and Semantic Scholar (40-60% PubMed)
    const pubmedPercentage = 0.4 + (random() * 0.2); // 40-60%
    const pubmed = Math.floor(total * pubmedPercentage);
    const semanticScholar = total - pubmed;

    // Calculate duplicates (around 15% of total)
    const duplicates = Math.floor(total * (0.13 + random() * 0.04)); // 13-17%

    // Calculate selected papers (10-25% of deduplicated papers)
    const deduplicatedTotal = total - duplicates;
    const selected = Math.floor(deduplicatedTotal * (0.1 + random() * 0.15)); // 10-25%

    const stats: DocumentCounts = {
      pubmed,
      semanticScholar,
      duplicates,
      selected,
      total
    };

    // Store the stats for this project
    this.projectStats.set(projectId, stats);
    return stats;
  }

  static updateStats(projectId: string, newStats: Partial<DocumentCounts>): DocumentCounts {
    const currentStats = this.projectStats.get(projectId) || this.generateInitialStats(projectId);
    const updatedStats = { ...currentStats, ...newStats };
    this.projectStats.set(projectId, updatedStats);
    return updatedStats;
  }

  static getStats(projectId: string): DocumentCounts {
    return this.projectStats.get(projectId) || this.generateInitialStats(projectId);
  }

  static clearStats(projectId: string): void {
    this.projectStats.delete(projectId);
  }
}

export default DocumentStats; 

File: backend/document_stats.py
import random
import json
import logging

logger = logging.getLogger(__name__)

class DocumentStats:
    @staticmethod
    def generate_stats(project_id: str) -> dict:
        try:
            logger.info(f"Generating stats for project: {project_id}")
            # Use the project_id (which is actually our query hash) as seed
            # Convert each character to its ASCII value and sum them
            seed = sum(ord(c) for c in project_id)
            random.seed()
            
            # Calculate total papers (800-1500)
            total = random.randint(800, 1500)
            
            # Split between PubMed and Semantic Scholar (40-60% PubMed)
            pubmed_percentage = 0.4 + (random.random() * 0.2)  # 40-60%
            pubmed = int(total * pubmed_percentage)
            semantic_scholar = total - pubmed
            
            # Calculate duplicates (around 15% of total)
            duplicates = int(total * (0.13 + random.random() * 0.04))  # 13-17%
            
            # Calculate selected papers (10-25% of deduplicated papers)
            deduplicated_total = total - duplicates
            selected = int(deduplicated_total * (0.1 + random.random() * 0.15))  # 10-25%
            
            stats = {
                "pubmed": pubmed,
                "semanticScholar": semantic_scholar,
                "duplicates": duplicates,
                "selected": selected,
                "total": total
            }
            logger.info(f"Generated stats: {stats}")
            return stats
        except Exception as e:
            logger.error(f"Error in generate_stats: {str(e)}")
            raise

    @staticmethod
    def generate_stats_for_query(query: str) -> dict:
        try:
            logger.info(f"Generating stats for query: {query}")
            # Use the query string itself as seed
            seed = sum(ord(c) for c in query)
            random.seed(seed)
            
            total = random.randint(800, 1500)
            duplicates = int(total * (0.13 + random.random() * 0.04))
            
            stats = {
                "total": total,
                "duplicates": duplicates
            }
            logger.info(f"Generated query stats: {stats}")
            return stats
        except Exception as e:
            logger.error(f"Error in generate_stats_for_query: {str(e)}")
            raise 